centrifuge/client.py

```diff
@@ -1,45 +1,92 @@
-import base64
 import asyncio
+import base64
+import contextlib
 import logging
+from contextlib import asynccontextmanager
 from dataclasses import dataclass
 from enum import Enum
-from typing import Any, Optional, Dict
-from typing import Callable, Awaitable
-from contextlib import asynccontextmanager
+from typing import TYPE_CHECKING, Any, Awaitable, Callable, Dict, Optional
+
 import websockets
+from websockets.protocol import State
 
 from centrifuge.codecs import _JsonCodec, _ProtobufCodec
-from centrifuge.codes import _DisconnectedCode, _ConnectingCode, _UnsubscribedCode, _SubscribingCode, _ErrorCode
-from centrifuge.contexts import PublicationContext, SubscribedContext, SubscribingContext, UnsubscribedContext, \
-    SubscriptionErrorContext, LeaveContext, JoinContext, SubscriptionTokenContext, DisconnectedContext, \
-    ConnectingContext, ConnectedContext, ErrorContext, ConnectionTokenContext, ServerLeaveContext, ServerJoinContext, \
-    ServerPublicationContext, ServerUnsubscribedContext, ServerSubscribedContext
-from centrifuge.exceptions import ClientDisconnected, ReplyError, CentrifugeException, Unauthorized, \
-    DuplicateSubscription, Timeout, SubscriptionUnsubscribed
+from centrifuge.codes import (
+    _ConnectingCode,
+    _DisconnectedCode,
+    _ErrorCode,
+    _SubscribingCode,
+    _UnsubscribedCode,
+)
+from centrifuge.contexts import (
+    ConnectedContext,
+    ConnectingContext,
+    ConnectionTokenContext,
+    DisconnectedContext,
+    ErrorContext,
+    JoinContext,
+    LeaveContext,
+    PublicationContext,
+    ServerJoinContext,
+    ServerLeaveContext,
+    ServerPublicationContext,
+    ServerSubscribedContext,
+    ServerUnsubscribedContext,
+    SubscribedContext,
+    SubscribingContext,
+    SubscriptionErrorContext,
+    SubscriptionTokenContext,
+    UnsubscribedContext,
+)
+from centrifuge.exceptions import (
+    CentrifugeError,
+    ClientDisconnectedError,
+    ClientTimeoutError,
+    DuplicateSubscriptionError,
+    ReplyError,
+    SubscriptionUnsubscribedError,
+    UnauthorizedError,
+)
 from centrifuge.handlers import _ConnectionEventHandler, _SubscriptionEventHandler
-from centrifuge.types import StreamPosition, PublishResult, BytesOrJSON, PresenceStatsResult, PresenceResult, \
-    HistoryResult, ClientInfo, RpcResult, Publication
-from centrifuge.utils import _backoff, _code_number, _code_message, _is_token_expired, _wait_for_future
-
-logger = logging.getLogger('centrifuge')
+from centrifuge.types import (
+    BytesOrJSON,
+    ClientInfo,
+    HistoryResult,
+    PresenceResult,
+    PresenceStatsResult,
+    Publication,
+    PublishResult,
+    RpcResult,
+    StreamPosition,
+)
+from centrifuge.utils import (
+    _backoff,
+    _code_message,
+    _code_number,
+    _is_token_expired,
+    _wait_for_future,
+)
+
+if TYPE_CHECKING:
+    from websockets.legacy.client import WebSocketClientProtocol
```

Comment:
Just noticed due to this import - looks like we are relying on some legacy client here.. Will be nice to find out what does it mean.

Reply:
https://websockets.readthedocs.io/en/stable/reference/asyncio/client.html#using-a-connection

https://github.com/python-websockets/websockets/blob/12.0/src/websockets/legacy/client.py#L56-L337

Reply:
Found this: https://websockets.readthedocs.io/en/stable/faq/misc.html#why-is-the-default-implementation-located-in-websockets-legacy

Reply:
yea, I found it too

Reply:
So looks like it's normal, will try to double-check maybe it's possible to import from another place - but seems that no refactoring required